lib/auth/password.go

```diff
@@ -354,7 +354,7 @@ func (s *AuthServer) changeUserSecondFactor(req ChangePasswordWithTokenRequest,
 		u2fRes := req.U2FRegisterResponse
 		reg, err := u2f.Register(u2fRes, *challenge, &u2f.Config{SkipAttestationVerify: true})
 		if err != nil {
-			return trace.Wrap(err)
+			return trace.AccessDenied(err.Error())
```

Comment:
Could you elaborate how this changes the behavior?
Also, if possible add a unit test for this to prevent regressions

Reply:
This is not a good place to return access denied error. It looks we need to convert u2f errors into our appropriate trace error so it can be serialized correctly. Right now this json looks like this

```json
{
  "message": "{ \"message\": \"u2f: xxxx\" \"}"
}
```

I think trace.Wrap(err.Error()) would do. No need to unit-test this.

Reply:
I updated my comment: https://github.com/gravitational/teleport/pull/3592#issue-404087266

The u2f is a 3rd party library and sends back a string based error: https://github.com/tstranex/u2f/blob/master/util.go#L115

And when it gets marshaled twice (goes through auth/apiserver, then web/apiserver), the first marshaled value doesn't get [unmarshaled](https://github.com/gravitational/teleport/blob/master/vendor/github.com/gravitational/trace/httplib.go#L78), b/c we never wrapped it in a trace defined error object.

So I wrapped the error message with the AccessDenied object. And I thought `u2f: untrusted facet id` is a type of AccessDenied error which gives us status code 403, which is more accurate than code 500.

Reply:
I replaced AccessDenied with BadParamter as Alexey suggested: https://github.com/gravitational/teleport/pull/3592/commits/303b3f484497feb9462ca83ba1fa0d72d9f8040a

Reply:
Thanks. Please leave a comment in the code explaining the nuance, so that a different person in the future doesn't decide to "clean up".

Reply:
revised: https://github.com/gravitational/teleport/pull/3592/commits/bcd67e0775c38337a4c693cf4e2fa4922d4a1074